### PR TITLE
[Fix] Preserve KDA summary factor-warp parity across work items

### LINKED ISSUE #1052: [Bug]: FROST KDA forward hangs with batch_invariant=True at T=16224
On B300, this forward call hangs with finite, all-zero inputs. It does not finish within 60 seconds. Changing only `T` from `16224` to `16384` completes successfully and returns exactly zero.

Reproduced with Python kernel sources from `develop` at `b2712de2da3832e9ad607867d22b2a1429c89cdc`, cuDNN Frontend native extension 1.29.0, backend 9.24.0, PyTorch 2.9.0 / CUDA 13.0, CuTe DSL 4.7.0, NVIDIA B300 (SM103).

Save as `repro.py` and run `timeout -k 5 60 python3 -u repro.py`:

```python
import torch
from cudnn.linear_attention import kimi_delta_attention

torch.cuda.set_device(0)
T, H, D = 16224, 128, 128
x = torch.zeros((T, H, D), device="cuda", dtype=torch.bfloat16)
beta = torch.ones((T, H), device="cuda", dtype=torch.bfloat16)
cu = torch.tensor([0, T], device="cuda", dtype=torch.int32)
a_log = torch.zeros(H, device="cuda")
dt_bias = torch.full((H, D), -5.5, device="cuda")
torch.cuda.synchronize()
print("Calling KDA", flush=True)
out = kimi_delta_attention(
    q=x, k=x, v=x, g=x, beta=beta, cu_seqlens=cu,
    a_log=a_log, dt_bias=dt_bias,
    safe_gate=True, gate_lower_bound=-5.0, batch_invariant=True,
    initial_state=None, output_final_state=False, plan_name="kda_frost",
)[0]
print("Forward returned; synchronizing", flush=True)
torch.cuda.synchronize()
assert torch.equal(out, torch.zeros_like(out))
print("PASS", flush=True)
```

Expected: the call completes with an all-zero output for either length.

Observed: `T=16224` prints `Calling KDA` and times out (exit 124); `T=16384` reaches `PASS` in a fresh process.

### PR DESCRIPTION
**Pull request by mdy666**
## Summary

Fixes #1052.

Keep the two summary factor warps assigned to their **cumulative** chunk parity across work items. Restarting at the same local parity after an odd-sized item swaps ring-slot ownership; independently advancing warps can then observe a stale barrier phase and hang.

The fix changes only the loop start in `super_mma_warp`, plus the corresponding warp-role descriptions. CG0, precision, formulas, launch grids and public APIs are unchanged. The affected path is the Blackwell FROST KDA summary kernel when one CTA processes multiple work items with an odd chunk count.

## Validation

B300 (SM103), PyTorch 2.9.0 / CUDA 13.0, cuDNN backend 9.24.0, CuTe DSL 4.7.0. BF16, safe gate with lower bound -5.

```bash
cd test/python
pytest -q -s -m L1 linear_attention/test_kda_summary_parity.py
```

- New regression: 2 passed. T=16224 exercises odd-sized pieces; T=16384 is the even-sized control. Checks forward and backward against the exact all-zero reference.
- Red check: the same T=16224 test on unmodified upstream times out after 60 seconds.
- Six nonzero synthetic cases forcing one summary CTA to reuse work items passed FP64-reference checks, covering odd/even lengths, ragged boundaries and optional state. Existing 2% forward / 4% backward limits were unchanged.
- The previously hanging 32K ragged case completes forward/backward with finite results. On an unaffected 32K single-sequence case, output and all seven gradients retain identical byte hashes before/after the fix.

No performance improvement is claimed; the failing workload previously did not complete. The broader validation and factor-only/CG0-only ablation are recorded in #1052.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved linear attention processing at odd-sized chunk boundaries.
  - Prevented skipped synchronization steps during cumulative processing.
  - Improved consistency between packed and standalone sequence results, including outputs, final states, and gradients.

- **Tests**
  - Added coverage for long sequences with odd chunk counts.
  - Verified packed results against standalone execution, including outputs, final states, and input and initial-state gradients.

### CODE REVIEW

**Comment by mdy666** [comment]
Could a maintainer please add `cat-bugfix`, `area:linear_attention`, and `orig-external`, and assign the required milestone/project? The merge-requirements check currently reports a missing milestone and project; I do not have permission to set these.

**Comment by YangXu1990uiuc** [comment]
**Codex bot review · model `gpt-6-astra`**

I approved the pending fork Style workflow and am requesting test CI for `69592bbf6c32` with the next standalone command. For future updates, the appropriate request is `@cudnn-ci-bot run frost` in its own comment; a maintainer may need to send it. I am also running the new odd/even regression explicitly on SM100, along with a baseline control.

**Comment by YangXu1990uiuc** [comment]
@cudnn-ci-bot run frost

**Review by YangXu1990uiuc — APPROVED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `69592bbf6c32` against base `0ed2710bc6db`.

The cumulative-parity fix makes sense: each factor warp keeps ownership of the same two-stage ring slot across odd-length work items. I independently reproduced the old-head hang and verified the fix on SM100. No P0 or high-risk P1 found; please address the test-selection gap below and the existing device-selection comment.

**[P2] Make the new regression survive both CI selection filters** ([source](https://github.com/NVIDIA/cudnn-frontend/blob/69592bbf6c3243cce2eceff225c810bdae608cbe/test/python/linear_attention/test_kda_summary_parity.py#L11))

These cases are currently deselected by two independent filters: pytest.ini defaults to -m L0 while the module is L1, and the FROST linear-attention invocation selects -k frost while neither the file nor the test name contains frost. Changing only the marker still leaves this hang regression untested. Please put it in the intended CI tier and give its node IDs the matching frost keyword, or explicitly include these node IDs in the job.

Evidence: Exact-head collection selected zero cases both with the default invocation and with -m "L0 or L1" -k frost; each reported two deselected tests and exit code 5. Explicitly selecting the file with -m "L0 or L1" ran and passed both cases. Attribution: The newly added regression does not enter the existing CI selection; the kernel correction itself passes the targeted reproduction.

- **[P2] Still open:** [Use the current CUDA device for the SM-count gate](https://github.com/NVIDIA/cudnn-frontend/pull/1064#discussion_r4011573264).

Validation: SM100 with 148 SMs, CuTe DSL 4.8, cuDNN 9.25.1: both new forward/backward zero-value regressions passed on 69592bbf. Using the same tests against develop 8cb4fc29 with isolated compilation caches, T=16384 passed and T=16224 hit a 70-second timeout after the compilation warm-up. A nonzero BF16 KDA forward at T=16384, H=D=128, batch_invariant=True produced byte-identical baseline/head output. CUDA Graph replay matched eager output, and changing V changed the result and still matched a fresh call. For that nonzero forward spot check, GPU medians were 2.965 ms baseline and 2.961 ms head. CPU call medians were 148 and 160 us, but individual samples overlapped substantially and drifted downward during both runs; this does not establish a CPU regression or a precise performance bound. Read the complete diff and traced the work-item handoff and ring phases; the prospective merge into develop 8cb4fc29 was clean. The specific fork Style run was approved and passed. I also started the matching-head FROST CI run; GPU suites are still running. Its current text-scan failure is in an unchanged convolution docstring. Before publication, develop advanced through #1070, confined to BSA code/docs/tests; the KDA source and its test dependencies are unchanged.

Limitations: The independent GPU checks used SM100. The author's SM103 and nonzero odd/ragged FP64 comparisons were read as supporting evidence but were not independently reproduced on SM103. Full FROST CI remains pending; the new cases require explicit selection until the P2 above is fixed.

Approved: no P0 or high-risk P1 found. Remaining findings stay with the owner; merge timing stays with the owner.

**Comment by YangXu1990uiuc** [comment]
**Codex bot review · model `gpt-6-astra`**

CI follow-up for `69592bbf6c32`: FROST linear on SM100 passed (1,574 passed, 169 skipped, 8 expected failures); SM100 GEMM and SM80/SM100/SM103/SM120 SDPA also passed. The remaining failures are Rubin SDPA tests outside the KDA change (three LSE-precision cases and one backend FP8-backward case), plus the text scan of an unchanged convolution comment. I have not established whether the Rubin failures also reproduce on the base.

The new odd/even KDA regression tests still need the CI-selection fixes noted in the review: this broad green linear run does not replace collecting those two tests.

**Comment by jhjpark** [comment]
@cudnn-ci-bot run frost

**Review by jhjpark — APPROVED** [review]

**Comment by jhjpark** [comment]
@cudnn-ci-bot run frost

**Review by YangXu1990uiuc — COMMENTED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `126b35165fcb` against base `2cf27d377e06`.

The follow-up is sound: the KDA correction is preserved, GDN2 now applies the same cumulative-parity invariant, and the new regression is actually selected by L0 plus the frost filter. All four new packed-versus-alone cases passed independently. No new approval blocker; existing approvals can stand. The small current-device test-gate issue remains at test_la.py:3454 after moving the regression.

- **[P2] Still open:** [Use the current CUDA device for the SM-count gate in the relocated test](https://github.com/NVIDIA/cudnn-frontend/pull/1064#discussion_r4011573264).

Validation: On SM100 with CuTe DSL 4.8 and cuDNN 9.25.1, default L0 collection with -k 'frost and odd_piece_boundary' selected all four new cases. Running that same selection passed 4/4 for GDN, KDA, GDN2 and GDP, including bitwise O, final state, all input gradients and initial-state gradients between packed and standalone sequences. Traced GDN2's cumulative chunk counter through both factor warps and the shared ring phases; the new first_chunk keeps global parity equal to chunk_parity across odd work items. Reused the earlier KDA red/green reproduction because the KDA correction itself is unchanged.

Limitations: Exact-head FROST CI pipeline 68139888 was already requested by a maintainer; I reused it. The linear lane is still running, and Style passed. This follow-up does not claim new target-SM103 or performance measurements.

Comment only: this PR already has an active approval; preserving existing approvals.
